File: tools/stream-migrate/testdata/chatmod/moderation.go
package main

import (
	"context"
	"log"

	stream "github.com/GetStream/stream-chat-go/v8"
)

// moderate covers the moderation surface: moderators, bans, mutes, flags and
// per-user blocking.
func (a *app) moderate(ctx context.Context) error {
	ch := a.client.Channel("messaging", "general")

	if _, err := ch.AddModerators(ctx, "user-alice"); err != nil {
		return err
	}
	if _, err := ch.DemoteModerators(ctx, "user-alice"); err != nil {
		return err
	}

	if _, err := a.client.BanUser(ctx, "user-bob", "user-alice",
		stream.BanWithReason("Spam"),
		stream.BanWithExpiration(60)); err != nil {
		return err
	}
	if _, err := a.client.UnBanUser(ctx, "user-bob"); err != nil {
		return err
	}
	if _, err := a.client.ShadowBan(ctx, "user-carol", "user-alice"); err != nil {
		return err
	}

	if _, err := a.client.MuteUser(ctx, "user-carol", "user-alice",
		stream.MuteWithExpiration(60)); err != nil {
		return err
	}
	if _, err := a.client.UnmuteUser(ctx, "user-carol", "user-alice"); err != nil {
		return err
	}

	banned, err := a.client.QueryBannedUsers(ctx, &stream.QueryBannedUsersOptions{
		QueryOption: &stream.QueryOption{
			Filter: map[string]interface{}{"channel_cid": "messaging:general"},
			Limit:  10,
		},
	})
	if err != nil {
		return err
	}
	log.Printf("%d bans", len(banned.Bans))

	return a.handleReports(ctx)
}

// handleReports is the flag and review workflow.
func (a *app) handleReports(ctx context.Context) error {
	if _, err := a.client.FlagMessage(ctx, "message-id", "user-alice"); err != nil {
		return err
	}
	if _, err := a.client.FlagUser(ctx, "user-carol", "user-alice"); err != nil {
		return err
	}

	flags, err := a.client.QueryMessageFlags(ctx, &stream.QueryOption{
		Filter: map[string]interface{}{"channel_cid": "messaging:general"},
	})
	if err != nil {
		return err
	}
	log.Printf("%d message flags", len(flags.Flags))

	reports, err := a.client.QueryFlagReports(ctx, &stream.QueryFlagReportsRequest{})
	if err != nil {
		return err
	}
	for _, report := range reports.FlagReports {
		if _, err := a.client.ReviewFlagReport(ctx, report.ID, &stream.ReviewFlagReportRequest{
			ReviewResult: "reviewed",
		}); err != nil {
			return err
		}
	}
	return nil
}

// blockPeer is the per-user block list, which is separate from banning.
func (a *app) blockPeer(ctx context.Context) error {
	if _, err := a.client.BlockUser(ctx, "user-carol", "user-alice"); err != nil {
		return err
	}
	blocked, err := a.client.GetBlockedUser(ctx, "user-alice")
	if err != nil {
		return err
	}
	log.Printf("%d blocked users", len(blocked.BlockedUsers))

	_, err = a.client.UnblockUser(ctx, "user-carol", "user-alice")
	return err
}

File: tools/fixtures/chatmod-python/moderation.py
"""Moderation surface of the representative integration: moderators, bans,
mutes, flags, per-user blocking, blocklists and push devices."""

from stream_chat import StreamChat


def moderate(chat: StreamChat) -> None:
    channel = chat.channel("messaging", "general")

    channel.add_moderators(["user-alice"])
    channel.demote_moderators(["user-alice"])

    chat.ban_user("user-bob", user_id="user-alice", reason="Spam", timeout=60)
    chat.unban_user("user-bob")
    chat.shadow_ban("user-carol", user_id="user-alice")
    chat.remove_shadow_ban("user-carol")

    chat.mute_user("user-carol", "user-alice", timeout=60)
    chat.unmute_user("user-carol", "user-alice")

    banned = chat.query_banned_users({"channel_cid": "messaging:general", "limit": 10})
    print(len(banned["bans"]), "bans")


def handle_reports(chat: StreamChat) -> None:
    chat.flag_message("message-id", user_id="user-alice")
    chat.flag_user("user-carol", user_id="user-alice")

    flags = chat.query_message_flags({"channel_cid": "messaging:general"})
    print(len(flags["flags"]), "message flags")

    chat.unflag_message("message-id", user_id="user-alice")


def block_peer(chat: StreamChat) -> None:
    chat.block_user("user-carol", "user-alice")
    blocked = chat.get_blocked_users("user-alice")
    print(len(blocked["blocks"]), "blocked users")
    chat.unblock_user("user-carol", "user-alice")


def configure_blocklists(chat: StreamChat) -> None:
    chat.create_blocklist("profanity", ["badword"])
    listing = chat.get_blocklist("profanity")
    print(listing["blocklist"]["name"], "has", len(listing["blocklist"]["words"]), "words")

    chat.update_blocklist("profanity", ["badword", "worse"])
    everything = chat.list_blocklists()
    print(len(everything["blocklists"]), "blocklists")
    chat.delete_blocklist("profanity")


def register_devices(chat: StreamChat) -> None:
    chat.add_device("device-token-123", "firebase", "user-alice")
    devices = chat.get_devices("user-alice")
    print(len(devices["devices"]), "devices")
    chat.delete_device("device-token-123", "user-alice")
